File: apps/evaluation-console/optimization_data.py
#!/usr/bin/env python3
"""Read-only access to the self-optimization loop's artifacts.

The optimization loop writes three artifact trees under the repo:

  optimization/results/<skill>/<iter>/   decision.json, summary.md, journal.jsonl
  optimization/runs/<skill>/<iter>/<scenario>/   screenshot.png, *-checks.json,
                                                 judge-verdicts.json, console.json, ...
  optimization/runs/<skill>/baseline/<scenario>/ the current-best render to diff against

This module turns those into clean JSON the evaluation console renders. It never
writes; the loop owns the files. Every path returned is repo-relative so the
server's ``/api/artifact`` endpoint can serve it.
"""

from __future__ import annotations

import json
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

# A non-terminal journal whose last event is older than this is a stalled /
# orphaned run, not "running" — generous enough not to false-stall an active
# iteration that is mid-step (the loop's steps can each take minutes).
RUNNING_MAX_AGE_S = 1800


def _is_fresh(ts_str: Any, max_age_s: int = RUNNING_MAX_AGE_S) -> bool:
    if not ts_str:
        return False
    try:
        ts = datetime.fromisoformat(str(ts_str))
    except ValueError:
        return False
    if ts.tzinfo is None:
        ts = ts.replace(tzinfo=timezone.utc)
    return (datetime.now(timezone.utc) - ts).total_seconds() <= max_age_s


REPO_ROOT = Path(__file__).resolve().parents[2]
RESULTS = REPO_ROOT / "optimization" / "results"
RUNS = REPO_ROOT / "optimization" / "runs"
SCENARIOS = REPO_ROOT / "optimization" / "scenarios"
SKILLS = REPO_ROOT / "skills"

# Per-scenario judge verdict -> who the visual panel preferred.
#   CANDIDATE = the new revision won (an improvement)
#   BASELINE  = the current-best won (a regression)
#   TIE       = no meaningful difference
WIN, LOSS, TIE = "CANDIDATE", "BASELINE", "TIE"


# --------------------------------------------------------------------------- #
# small helpers
# --------------------------------------------------------------------------- #
def _rel(path: Path) -> str:
    try:
        return str(path.resolve().relative_to(REPO_ROOT))
    except ValueError:
        return str(path)


def _read_json(path: Path) -> Any:
    try:
        with path.open("r", encoding="utf-8") as fh:
            return json.load(fh)
    except Exception:
        return None


def _iter_key(name: str) -> tuple[int, str]:
    """Sort 'baseline' first, then numeric iterations ascending."""
    if name == "baseline":
        return (-1, name)
    m = re.match(r"(\d+)", name)
    return (int(m.group(1)) if m else 10_000, name)


def _scenario_label(dirname: str) -> tuple[str, str]:
    """'eval-006-grand-canyon-south-rim' -> ('eval-006', 'grand canyon south rim')."""
    m = re.match(r"(eval-\d+)-(.*)$", dirname)
    if not m:
        return dirname, dirname.replace("-", " ")
    return m.group(1), m.group(2).replace("-", " ")


def _screenshots(bundle: Path) -> list[str]:
    if not bundle.is_dir():
        return []
    shots = sorted(bundle.glob("screenshot*.png"), key=lambda p: p.name)
    return [_rel(p) for p in shots]


# --------------------------------------------------------------------------- #
# summary.md scores
# --------------------------------------------------------------------------- #
_SCORE_PATTERNS = {
    "programmatic": r"\*\*Programmatic Correctness:\*\*\s*([\d.]+)%",
    "api": r"\*\*API Accuracy:\*\*\s*([\d.]+)%",
    "visual_win_rate": r"\*\*Visual Win Rate:\*\*\s*([\d.]+)%",
    "coverage_delta": r"\*\*Coverage Delta:\*\*\s*([\-\d.]+)%",
}


def _parse_scores(summary_md: Path) -> dict[str, float | None]:
    out: dict[str, float | None] = {k: None for k in _SCORE_PATTERNS}
    if not summary_md.is_file():
        return out
    text = summary_md.read_text(encoding="utf-8", errors="replace")
    for key, pat in _SCORE_PATTERNS.items():
        m = re.search(pat, text)
        if m:
            try:
                out[key] = float(m.group(1))
            except ValueError:
                pass
    return out


# --------------------------------------------------------------------------- #
# iteration-level reads
# --------------------------------------------------------------------------- #
def _iteration_dirs(skill: str) -> list[str]:
    base = RESULTS / skill
    if not base.is_dir():
        return []
    names = [p.name for p in base.iterdir() if p.is_dir()]
    return sorted(names, key=_iter_key)


def _decision(skill: str, it: str) -> dict[str, Any]:
    return _read_json(RESULTS / skill / it / "decision.json") or {}


def _journal(skill: str, it: str) -> list[dict[str, Any]]:
    path = RESULTS / skill / it / "journal.jsonl"
    if not path.is_file():
        return []
    events: list[dict[str, Any]] = []
    for line in path.read_text(encoding="utf-8", errors="replace").splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            events.append(json.loads(line))
        except json.JSONDecodeError:
            continue
    return events


def iteration_summary(skill: str, it: str) -> dict[str, Any]:
    """Lightweight per-iteration record for the history strip + trends."""
    decision = _decision(skill, it)
    scores = _parse_scores(RESULTS / skill / it / "summary.md")
    journal = _journal(skill, it)
    started = next((e.get("timestamp_utc") for e in journal if e.get("event") == "iteration_started"), None)
    finished = next(
        (e.get("timestamp_utc") for e in reversed(journal) if e.get("event") == "iteration_completed"),
        None,
    )
    counts = decision.get("counts") or {}
    events = {e.get("event") for e in journal}
    if it == "baseline":
        status = "baseline"
    elif "iteration_completed" in events:
        status = "completed"
    elif "iteration_failed" in events:
        status = "failed"
    elif journal:
        # Non-terminal journal: "running" only if its last event is recent;
        # otherwise it is a stalled/orphaned run from a previous session.
        last_ts = journal[-1].get("timestamp_utc")
        status = "running" if _is_fresh(last_ts) else "stalled"
    else:
        status = "empty"
    failed_step = next(
        (e.get("step") for e in reversed(journal) if e.get("event") == "iteration_failed"),
        None,
    )
    return {
        "iteration": it,
        "is_baseline": it == "baseline",
        "status": status,
        "failed_step": failed_step,
        "decision": decision.get("decision"),
        "rule_fired": decision.get("rule_fired"),
        "rationale": decision.get("rationale"),
        "counts": {
            "wins": counts.get("wins", 0),
            "losses": counts.get("losses", 0),
            "ties": counts.get("ties", 0),
            "critical_failures": counts.get("critical_failures", 0),
            "check_failures": counts.get("check_failures", 0),
        },
        "scores": scores,
        "started_utc": started,
        "finished_utc": finished,
        "has_runs": (RUNS / skill / it).is_dir(),
    }


# --------------------------------------------------------------------------- #
# scenario-level reads (the candidate-vs-baseline evidence)
# --------------------------------------------------------------------------- #
def _checks_summary(bundle: Path) -> dict[str, int]:
    data = _read_json(bundle / "programmatic-checks.json") or {}
    checks = data.get("checks") or []
    passed = sum(1 for c in checks if c.get("result") == "pass")
    crit_failed = sum(1 for c in checks if c.get("result") == "fail" and c.get("critical"))
    return {"passed": passed, "total": len(checks), "critical_failed": crit_failed}


def _console_errors(bundle: Path) -> int:
    data = _read_json(bundle / "console.json") or {}
    return len(data.get("errors") or [])


def _scenario_detail(skill: str, it: str, scenario_dir: Path) -> dict[str, Any]:
    scenario_id, label = _scenario_label(scenario_dir.name)
    verdicts = _read_json(scenario_dir / "judge-verdicts.json") or {}
    baseline_bundle = RUNS / skill / "baseline" / scenario_dir.name
    individual = [
        {
            "seed": v.get("seed"),
            "verdict": v.get("verdict"),
            "rationale": v.get("rationale"),
            "judge_index": v.get("judge_index"),
        }
        for v in (verdicts.get("individual_verdicts") or [])
    ]
    return {
        "scenario_id": scenario_id,
        "label": label,
        "dir": scenario_dir.name,
        "verdict": verdicts.get("verdict"),
        "majority_count": verdicts.get("majority_count"),
        "judge_unavailable": bool(verdicts.get("judge_unavailable")),
        "individual_verdicts": individual,
        "checks": _checks_summary(scenario_dir),
        "console_errors": _console_errors(scenario_dir),
        "candidate_screenshots": _screenshots(scenario_dir),
        "baseline_screenshots": _screenshots(baseline_bundle),
    }


def iteration_detail(skill: str, it: str) -> dict[str, Any]:
    """Full evidence for one iteration: decision + scores + per-scenario diffs."""
    summary = iteration_summary(skill, it)
    runs_dir = RUNS / skill / it
    scenarios: list[dict[str, Any]] = []
    if runs_dir.is_dir():
        for sd in sorted(runs_dir.iterdir(), key=lambda p: p.name):
            if sd.is_dir() and sd.name.startswith("eval-"):
                scenarios.append(_scenario_detail(skill, it, sd))
    summary["scenarios"] = scenarios
    summary["journal"] = _journal(skill, it)
    return summary


# --------------------------------------------------------------------------- #
# skill-level aggregation
# --------------------------------------------------------------------------- #
def _live_skill_md_meta(skill: str) -> dict[str, Any]:
    md = SKILLS / skill / "SKILL.md"
    if not md.is_file():
        return {"exists": False}
    text = md.read_text(encoding="utf-8", errors="replace")
    return {"exists": True, "bytes": len(text.encode("utf-8")), "lines": text.count("\n") + 1}


def skill_overview(skill: str) -> dict[str, Any]:
    iters = _iteration_dirs(skill)
    history = [iteration_summary(skill, it) for it in iters]
    non_baseline = [h for h in history if not h["is_baseline"]]
    kept = sum(1 for h in non_baseline if h["decision"] == "KEEP")
    rejected = sum(1 for h in non_baseline if h["decision"] == "REJECT")
    # "Latest" is the most recent decisioned iteration, not a half-finished attempt.
    decisioned = [h for h in non_baseline if h["decision"]]
    latest = decisioned[-1] if decisioned else (non_baseline[-1] if non_baseline else None)
    running = [h for h in non_baseline if h["status"] == "running"]
    return {
        "skill": skill,
        "iteration_count": len(non_baseline),
        "kept": kept,
        "rejected": rejected,
        "latest": latest,
        "running": bool(running),
        "history": history,
        "skill_md": _live_skill_md_meta(skill),
    }


def list_skills() -> list[dict[str, Any]]:
    if not RESULTS.is_dir():
        return []
    skills = sorted(p.name for p in RESULTS.iterdir() if p.is_dir() and (RESULTS / p.name).is_dir())
    return [skill_overview(s) for s in skills if _iteration_dirs(s)]


# --------------------------------------------------------------------------- #
# live-run detection (the loop streams into journal.jsonl as it runs)
# --------------------------------------------------------------------------- #
_TERMINAL_EVENTS = {"iteration_completed", "iteration_failed", "baseline_check_completed"}


def active_runs() -> list[dict[str, Any]]:
    """Iterations whose journal has started but reached no terminal event.

    Note: a stale journal (process died without a terminal event) also lands
    here. True "is a process running right now" is owned by the server's
    run-control registry; this is the disk-level signal only.
    """
    live: list[dict[str, Any]] = []
    if not RESULTS.is_dir():
        return live
    for skill_dir in RESULTS.iterdir():
        if not skill_dir.is_dir():
            continue
        for it in _iteration_dirs(skill_dir.name):
            journal = _journal(skill_dir.name, it)
            if not journal:
                continue
            last = journal[-1]
            if last.get("event") in _TERMINAL_EVENTS:
                continue
            live.append(
                {
                    "skill": skill_dir.name,
                    "iteration": it,
                    "last_event": last.get("event"),
                    "last_step": last.get("step"),
                    "last_ts": last.get("timestamp_utc"),
                }
            )
    return live


if __name__ == "__main__":  # quick smoke check
    import sys

    out = {"skills": [s["skill"] for s in list_skills()], "active": active_runs()}
    json.dump(out, sys.stdout, indent=2)
    print()

File: packages/eval/src/console/optimizationData.ts
/**
 * Read-only access to the self-optimization loop's artifacts
 * (results/runs/scenarios trees) for the evaluation console.
 */
import * as fs from "node:fs";
import * as path from "node:path";
import { readJson, readJsonOrNull, readJsonl } from "../lib/json.js";
import { fromRepoRoot, globFiles, listDirs, repoRelative } from "../lib/paths.js";
import { parseTs } from "../lib/format.js";

export const WIN = "CANDIDATE";
export const LOSS = "BASELINE";
export const TIE = "TIE";

const resultsRoot = () => fromRepoRoot("optimization", "results");
const runsRoot = () => fromRepoRoot("optimization", "runs");
const scenariosRoot = () => fromRepoRoot("optimization", "scenarios");
const skillsRoot = () => fromRepoRoot("skills");
const candidatesRoot = () => fromRepoRoot("optimization", "candidates");
const historyRoot = () => fromRepoRoot("optimization", "history");

/**
 * Human-review and promotion state of a KEEP candidate:
 *  - "staged":   PROMOTED-PENDING.md awaits a Decide review
 *  - "approved": Decide approved the candidate for the separate Promote gate
 *  - "rejected": Decide rejected the candidate; skills/ remains untouched
 *  - "promoted": updateCurrentBest ran (backup exists), SKILL.md was updated
 *  - "unknown":  KEEP with neither marker (should not happen; stay honest)
 */
export function classifyPromotionState(input: {
  pending: boolean;
  promoted: boolean;
  reviewDecision: unknown;
}): string {
  if (input.promoted) return "promoted";
  if (input.pending && input.reviewDecision === "approve") return "approved";
  if (input.pending && input.reviewDecision === "reject") return "rejected";
  if (input.pending) return "staged";
  return "unknown";
}

export function promotionState(skill: string, iteration: string): string {
  const candidateDir = path.join(candidatesRoot(), skill, iteration);
  const review = readJsonOrNull(path.join(candidateDir, "candidate-review.json"));
  return classifyPromotionState({
    pending: fs.existsSync(path.join(candidateDir, "PROMOTED-PENDING.md")),
    promoted: fs.existsSync(path.join(historyRoot(), skill, `iteration-${iteration}`, "current-best-before.md")),
    reviewDecision: review?.decision,
  });
}

/** A non-terminal journal older than this is stalled, not running. */
export function isFresh(tsValue: unknown, maxAgeSeconds: number): boolean {
  const ts = parseTs(tsValue);
  if (ts === null) return false;
  return (Date.now() - ts.getTime()) / 1000 <= maxAgeSeconds;
}

function iterSortKey(name: string): [number, string] {
  if (name === "baseline") return [-1, name];
  const match = /^(\d+)/.exec(name);
  return [match ? Number(match[1]) : 10_000, name];
}

export function scenarioLabel(dirname: string): [string, string] {
  const match = /^(eval-\d+)-(.*)$/.exec(dirname);
  if (!match) return [dirname, dirname.replace(/-/g, " ")];
  return [match[1], match[2].replace(/-/g, " ")];
}

function screenshots(bundle: string): string[] {
  return globFiles(bundle, "screenshot", ".png").map(repoRelative);
}

// ---------------------------------------------------------------------------
// summary.md scores
// ---------------------------------------------------------------------------
const SCORE_PATTERNS: Record<string, RegExp> = {
  programmatic: /\*\*Programmatic Correctness:\*\*\s*([\d.]+)%/,
  api: /\*\*API Accuracy:\*\*\s*([\d.]+)%/,
  visual_win_rate: /\*\*Visual Win Rate:\*\*\s*([\d.]+)%/,
  coverage_delta: /\*\*Coverage Delta:\*\*\s*([-\d.]+)%/,
};

function parseScores(summaryMd: string): Record<string, number | null> {
  const out: Record<string, number | null> = Object.fromEntries(Object.keys(SCORE_PATTERNS).map((key) => [key, null]));
  if (!fs.existsSync(summaryMd)) return out;
  const text = fs.readFileSync(summaryMd, "utf-8");
  for (const [key, pattern] of Object.entries(SCORE_PATTERNS)) {
    const match = pattern.exec(text);
    if (match) {
      const value = Number(match[1]);
      if (!Number.isNaN(value)) out[key] = value;
    }
  }
  return out;
}

// ---------------------------------------------------------------------------
// iteration-level reads
// ---------------------------------------------------------------------------
export function iterationDirs(skill: string): string[] {
  return listDirs(path.join(resultsRoot(), skill)).sort((a, b) => {
    const [na, sa] = iterSortKey(a);
    const [nb, sb] = iterSortKey(b);
    return na - nb || (sa < sb ? -1 : sa > sb ? 1 : 0);
  });
}

export function journalFor(skill: string, iteration: string): Array<Record<string, any>> {
  return readJsonl(path.join(resultsRoot(), skill, iteration, "journal.jsonl"));
}

/**
 * A baseline can be retried in place, so its journal may contain several
 * attempts. Only events after the latest start describe the current state.
 * Candidate iterations are immutable in normal operation, but applying the
 * same rule there also makes interrupted/retried development runs honest.
 */
export function currentJournalAttempt(
  iteration: string,
  journal: Array<Record<string, any>>,
): Array<Record<string, any>> {
  const startEvent = iteration === "baseline" ? "baseline_check_started" : "iteration_started";
  for (let index = journal.length - 1; index >= 0; index -= 1) {
    if (journal[index]?.event === startEvent) return journal.slice(index);
  }
  return journal;
}

function eventError(event: Record<string, any> | undefined): string | null {
  if (!event) return null;
  if (typeof event.error === "string" && event.error.trim()) return event.error.trim();
  if (event.result && typeof event.result.error === "string" && event.result.error.trim()) {
    return event.result.error.trim();
  }
  return null;
}

/** Pure journal-to-lifecycle projection shared by the API and regression tests. */
export function journalLifecycle(
  iteration: string,
  journalIn: Array<Record<string, any>>,
  runningMaxAgeSeconds: number,
): Record<string, any> {
  const journal = currentJournalAttempt(iteration, journalIn);
  const started =
    journal.find((e) => e.event === (iteration === "baseline" ? "baseline_check_started" : "iteration_started"))
      ?.timestamp_utc ?? null;
  const failedEvent = [...journal]
    .reverse()
    .find((e) => e.event === "iteration_failed" || String(e.event ?? "").endsWith("_failed"));
  const completedEvent = [...journal]
    .reverse()
    .find(
      (e) =>
        e.event === "iteration_completed" ||
        e.event === "baseline_check_completed" ||
        e.event === "baseline_browser_eval_completed",
    );
  const events = new Set(journal.map((e) => e.event));

  let status: string;
  if (failedEvent) status = "failed";
  else if (
    iteration === "baseline" &&
    (events.has("baseline_check_completed") || events.has("baseline_browser_eval_completed"))
  ) {
    status = "baseline";
  } else if (events.has("iteration_completed")) status = "completed";
  else if (journal.length) {
    status = isFresh(journal[journal.length - 1]?.timestamp_utc, runningMaxAgeSeconds) ? "running" : "stalled";
  } else status = "empty";

  return {
    journal,
    status,
    started_utc: started,
    finished_utc: (failedEvent ?? completedEvent)?.timestamp_utc ?? null,
    failed_step:
      failedEvent?.step ??
      (typeof failedEvent?.event === "string" ? failedEvent.event.replace(/_failed$/, "") : null),
    error: eventError(failedEvent),
  };
}

export function iterationSummary(skill: string, iteration: string, runningMaxAgeSeconds: number): Record<string, any> {
  const decision = readJsonOrNull(path.join(resultsRoot(), skill, iteration, "decision.json")) ?? {};
  const scores = parseScores(path.join(resultsRoot(), skill, iteration, "summary.md"));
  const lifecycle = journalLifecycle(iteration, journalFor(skill, iteration), runningMaxAgeSeconds);
  const counts = decision.counts ?? {};

  return {
    iteration,
    is_baseline: iteration === "baseline",
    status: lifecycle.status,
    failed_step: lifecycle.failed_step,
    error: lifecycle.error,
    decision: decision.decision ?? null,
    rule_fired: decision.rule_fired ?? null,
    rationale: decision.rationale ?? null,
    promotion: decision.decision === "KEEP" ? promotionState(skill, iteration) : null,
    counts: {
      wins: counts.wins ?? 0,
      losses: counts.losses ?? 0,
      ties: counts.ties ?? 0,
      critical_failures: counts.critical_failures ?? 0,
      check_failures: counts.check_failures ?? 0,
    },
    scores,
    started_utc: lifecycle.started_utc,
    finished_utc: lifecycle.finished_utc,
    has_runs: fs.existsSync(path.join(runsRoot(), skill, iteration)),
  };
}

// ---------------------------------------------------------------------------
// scenario-level reads (candidate-vs-baseline evidence)
// ---------------------------------------------------------------------------
function checksSummary(bundle: string): Record<string, number> {
  const data = readJsonOrNull(path.join(bundle, "programmatic-checks.json")) ?? {};
  const checks: Array<Record<string, any>> = data.checks ?? [];
  return {
    passed: checks.filter((c) => c.result === "pass").length,
    total: checks.length,
    critical_failed: checks.filter((c) => c.result === "fail" && c.critical).length,
  };
}

function consoleErrors(bundle: string): number {
  const data = readJsonOrNull(path.join(bundle, "console.json")) ?? {};
  return (data.errors ?? []).length;
}

function scenarioDetail(skill: string, iteration: string, scenarioDir: string): Record<string, any> {
  const dirname = path.basename(scenarioDir);
  const [scenarioId, label] = scenarioLabel(dirname);
  const verdicts = readJsonOrNull(path.join(scenarioDir, "judge-verdicts.json")) ?? {};
  const baselineBundle = path.join(runsRoot(), skill, "baseline", dirname);
  return {
    scenario_id: scenarioId,
    label,
    dir: dirname,
    verdict: verdicts.verdict ?? null,
    majority_count: verdicts.majority_count ?? null,
    judge_unavailable: Boolean(verdicts.judge_unavailable),
    individual_verdicts: (verdicts.individual_verdicts ?? []).map((v: any) => ({
      seed: v.seed ?? null,
      verdict: v.verdict ?? null,
      rationale: v.rationale ?? null,
      judge_index: v.judge_index ?? null,
    })),
    checks: checksSummary(scenarioDir),
    console_errors: consoleErrors(scenarioDir),
    candidate_screenshots: screenshots(scenarioDir),
    baseline_screenshots: screenshots(baselineBundle),
  };
}

export function iterationDetail(skill: string, iteration: string, runningMaxAgeSeconds: number): Record<string, any> {
  const summary = iterationSummary(skill, iteration, runningMaxAgeSeconds);
  const runsDir = path.join(runsRoot(), skill, iteration);
  const scenarios: Array<Record<string, any>> = [];
  for (const name of listDirs(runsDir)) {
    if (name.startsWith("eval-")) scenarios.push(scenarioDetail(skill, iteration, path.join(runsDir, name)));
  }
  summary.scenarios = scenarios;
  summary.journal = currentJournalAttempt(iteration, journalFor(skill, iteration));
  // Proposer seed provenance: was this candidate seeded by a human-confirmed
  // scorecard focus (explicit decision path), or machine-initiated?
  const proposerMeta = readJsonOrNull(path.join(candidatesRoot(), skill, iteration, "proposer-metadata.json"));
  summary.proposer_seed = proposerMeta
    ? { explicit: proposerMeta.explicit_seed ?? null, decision_path: proposerMeta.decision_path ?? null }
    : null;
  return summary;
}

// ---------------------------------------------------------------------------
// skill-level aggregation
// ---------------------------------------------------------------------------
function skillMdMeta(skill: string): Record<string, any> {
  const md = path.join(skillsRoot(), skill, "SKILL.md");
  if (!fs.existsSync(md)) return { exists: false };
  const text = fs.readFileSync(md, "utf-8");
  return { exists: true, bytes: Buffer.byteLength(text, "utf-8"), lines: text.split("\n").length };
}

export function skillOverview(skill: string, runningMaxAgeSeconds: number): Record<string, any> {
  const iters = iterationDirs(skill);
  const history = iters.map((iteration) => iterationSummary(skill, iteration, runningMaxAgeSeconds));
  const nonBaseline = history.filter((h) => !h.is_baseline);
  const kept = nonBaseline.filter((h) => h.decision === "KEEP").length;
  const rejected = nonBaseline.filter((h) => h.decision === "REJECT").length;
  const decisioned = nonBaseline.filter((h) => h.decision);
  const latest = decisioned.length ? decisioned[decisioned.length - 1] : nonBaseline.length ? nonBaseline[nonBaseline.length - 1] : null;
  return {
    skill,
    iteration_count: nonBaseline.length,
    kept,
    rejected,
    latest,
    running: history.some((h) => h.status === "running"),
    history,
    skill_md: skillMdMeta(skill),
  };
}

export function listSkills(runningMaxAgeSeconds: number): Array<Record<string, any>> {
  if (!fs.existsSync(resultsRoot())) return [];
  return listDirs(resultsRoot())
    .filter((skill) => iterationDirs(skill).length)
    .map((skill) => skillOverview(skill, runningMaxAgeSeconds));
}

// ---------------------------------------------------------------------------
// live-run detection (disk-level signal)
// ---------------------------------------------------------------------------
const TERMINAL_EVENTS = new Set([
  "iteration_completed",
  "iteration_failed",
  "baseline_check_completed",
  "baseline_check_failed",
  "baseline_generation_failed",
  "baseline_browser_eval_completed",
  "baseline_browser_eval_failed",
]);

export function activeRuns(): Array<Record<string, any>> {
  const live: Array<Record<string, any>> = [];
  if (!fs.existsSync(resultsRoot())) return live;
  for (const skill of listDirs(resultsRoot())) {
    for (const iteration of iterationDirs(skill)) {
      const journal = journalFor(skill, iteration);
      if (!journal.length) continue;
      const last = journal[journal.length - 1];
      if (TERMINAL_EVENTS.has(last.event)) continue;
      live.push({
        skill,
        iteration,
        last_event: last.event ?? null,
        last_step: last.step ?? null,
        last_ts: last.timestamp_utc ?? null,
      });
    }
  }
  return live;
}
